ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -131,55 +176,58 @@ class DatasetToolArgs:
     )
 
     def __post_init__(self):
-        assert (
-            not self.upload_split or self.dataset_split
-        ), "Must specify dataset_split when using upload_split"
+        if not self.upload_subset:
+            self.upload_subset = self.dataset_subset
+        if self.dataset_split and not self.upload_split:
+            self.upload_split = self.dataset_split
 
 
 def main(args: DatasetToolArgs):
     ds_name = args.dataset_name
     print(f'Loading dataset "{ds_name}" for task {args.task}')
-    data_dict: datasets.DatasetDict = datasets.load_dataset(
+    ds: datasets.DatasetDict = datasets.load_dataset(
         ds_name, args.dataset_subset, split=args.dataset_split
     )
-    if args.dataset_split:
-        data_dict = datasets.DatasetDict(**{args.dataset_split: data_dict})
 
-    for split, ds_split in data_dict.items():
-        print(f'Processing split "{split}"...')
-        if args.shuffle:
-            ds_split = ds_split.shuffle(seed=args.shuffle_seed)
-        if args.num_samples:
-            ds_split = ds_split.select(range(args.num_samples))
-        data_dict[split] = args.task.map_split(ds_split, args.num_workers)
+    if isinstance(ds, datasets.Dataset):
+        ds = datasets.DatasetDict({args.upload_split: ds})
+
+    if len(ds) > 1 and args.upload_split:
+        raise ValueError("Cannot upload multiple splits to a single split")
 
-    token = args.token or os.environ.get("HF_TOKEN")
     hub_args: Dict[str, Any] = {
-        "config_name": args.dataset_subset or "default",
-        "token": token,
+        "config_name": args.upload_subset if args.upload_subset else "default",
+        "token": args.token or os.environ.get("HF_TOKEN"),
         "revision": args.upload_branch,
         "private": args.private,
-        "split": args.upload_split,
     }
     if args.num_shards is not None:
-        hub_args["num_shards"] = {split: args.num_shards for split in data_dict.keys()}
+        hub_args["num_shards"] = args.num_shards
 
-    try:
-        if args.dataset_split:
-            data_dict[args.dataset_split].push_to_hub(
-                args.upload_name, split=args.dataset_split, **hub_args
-            )
-        else:
-            data_dict.push_to_hub(args.upload_name, **hub_args)
-    except Exception as e:
-        print(f"Failed to push to hub: {e}")
-
-        # If the push fails or upload_name is not specified, save the data locally.
-        for split in data_dict.keys():
-            output_name = f"{split}-00000-of-00001.parquet"
-            data_dict[split].to_parquet(output_name)
+    for split, ds_split in ds.items():
```

Comment:
The sequential upload doesn't work if you change the data format (e.g. add a column). That's why we kept both options.

Reply:
You can create a new dataset to push to and rename it later, but that will completely destroy the history.

Reply:
I think this is still unresolved.